compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -5902,114 +5902,350 @@ OMR::Power::TreeEvaluator::generateHelperBranchAndLinkInstruction(
-   bool stopUsingCopyReg1, stopUsingCopyReg2 = false, stopUsingCopyReg3 = false;
+   bool arrayCheckNeeded;
 
-   stopUsingCopyReg1 = TR::TreeEvaluator::stopUsingCopyReg(dstAddrNode, dstAddrReg, cg);
+   // IL tree structure depends on whether or not it's been determined that a runtime arrayCHK is needed:
+   // if node has four children (i.e.: object base address and offset are separate), need array check
+   // if node three children (i.e.: object base address and offset have already been added together), don't need array check
+   if (node->getNumChildren() == 4)
+      {
+      arrayCheckNeeded = true;
+
+      dstBaseAddrNode = node->getChild(0);
+      dstOffsetNode = node->getChild(1);
+      dstAddrNode = NULL;
+      lengthNode = node->getChild(2);
+      valueNode = node->getChild(3);
+      }
+   else //i.e.: node->getNumChildren() == 3
+      {
+      arrayCheckNeeded = false;
+
+      dstBaseAddrNode = NULL;
+      dstOffsetNode = NULL;
+      dstAddrNode = node->getChild(0);
+      lengthNode = node->getChild(1);
+      valueNode = node->getChild(2);
+      }
+
+   TR::Register         *dstBaseAddrReg, *dstOffsetReg, *dstAddrReg, *lengthReg, *valueReg;
+
+   // if the offset is a constant value less than 16 bits, then we dont need a separate register for it
+   bool useOffsetAsImmVal = dstOffsetNode && dstOffsetNode->getOpCode().isLoadConst() &&
+                            (dstOffsetNode->getConstValue() >= LOWER_IMMED) && (dstOffsetNode->getConstValue() <= UPPER_IMMED);
+
+   bool stopUsingCopyRegBase = dstBaseAddrNode ? TR::TreeEvaluator::stopUsingCopyReg(dstBaseAddrNode, dstBaseAddrReg, cg) : false;
+   bool stopUsingCopyRegAddr = dstAddrNode ? TR::TreeEvaluator::stopUsingCopyReg(dstAddrNode, dstAddrReg, cg) : false ;
+
+   bool stopUsingCopyRegOffset, stopUsingCopyRegLen, stopUsingCopyRegVal;
+
+   //dstOffsetNode (type: long)
+   if (dstOffsetNode && !useOffsetAsImmVal) //only want to allocate a register for dstOffset if we're using it for the array check AND it isn't a constant
+      {
+      if (cg->comp()->target().is32Bit()) //on 32-bit systems, need to grab the lower 32 bits of the offset from the register pair
+         {
+         dstOffsetReg = cg->evaluate(dstOffsetNode);
+         TR::Register *offsetcopyReg = cg->allocateRegister();
+         generateTrg1Src1Instruction(cg, TR::InstOpCode::mr, dstOffsetNode, offsetcopyReg, dstOffsetReg->getLowOrder());
 
+         dstOffsetReg = offsetcopyReg;
+         stopUsingCopyRegOffset = true;
+         }
+      else
+         {
+         stopUsingCopyRegOffset = TR::TreeEvaluator::stopUsingCopyReg(dstOffsetNode, dstOffsetReg, cg);
+         }
+      }
+   else
+      {
+      stopUsingCopyRegOffset = false;
+      }
+
+   //lengthNode (type: long)
    lengthReg = cg->evaluate(lengthNode);
-   if (!cg->canClobberNodesRegister(lengthNode))
+   if (cg->comp()->target().is32Bit() || !cg->canClobberNodesRegister(lengthNode))
```

Comment:
same comment as above.  suggesting to program it for better readability, something like:
```
if (!canClobber...)
   {
   32bit/64bit handling;
   }
else
   {
   32bit/64bit handling;
   }
```

Reply:
I believe the `refCount > 1` check is already included in `!cg->canClobberNodesRegister()`, so is it necessary to add the ref count condition separately?

Reply:
no.  just for better readability then.